src/Microsoft.Net.Http.Headers/MediaTypeHeaderValue.cs

```diff
@@ -170,55 +212,148 @@ public IList<NameValueHeaderValue> Parameters
+        /// </example>
+        /// <remarks>See <see href="https://tools.ietf.org/html/rfc6838#section-4.2"/> for more details on the type.</remarks>
         public StringSegment Type
         {
             get
             {
-                return _mediaType.Subsegment(0, _mediaType.IndexOf('/'));
+                return _mediaType.Subsegment(0, _mediaType.IndexOf(ForwardSlashCharacter));
             }
         }
 
+        /// <summary>
+        /// Gets the subtype of the <see cref="MediaTypeHeaderValue"/>.
+        /// </summary>
+        /// <example>
+        /// For the media type <c>"application/vnd.example+json"</c>, the property gives the value
+        /// <c>"vnd.example+json"</c>.
+        /// </example>
+        /// <remarks>See <see href="https://tools.ietf.org/html/rfc6838#section-4.2"/> for more details on the subtype.</remarks>
         public StringSegment SubType
         {
             get
             {
-                return _mediaType.Subsegment(_mediaType.IndexOf('/') + 1);
+                return _mediaType.Subsegment(_mediaType.IndexOf(ForwardSlashCharacter) + 1);
             }
         }
 
         /// <summary>
-        /// MediaType = "*/*"
+        /// Gets subtype of the <see cref="MediaTypeHeaderValue"/>, excluding any structured syntax suffix. Returns <see cref="StringSegment.Empty"/>
+        /// if there is no subtype without suffix.
         /// </summary>
-        public bool MatchesAllTypes
+        /// <example>
+        /// For the media type <c>"application/vnd.example+json"</c>, the property gives the value
+        /// <c>"vnd.example"</c>.
+        /// </example>
+        public StringSegment SubTypeWithoutSuffix
         {
             get
             {
-                return MediaType.Equals("*/*", StringComparison.Ordinal);
+                var startOfSuffix = _mediaType.LastIndexOf(PlusCharacter);
+                var startOfSubType = _mediaType.IndexOf(ForwardSlashCharacter) + 1;
+                if (startOfSuffix == -1)
+                {
+                    return _mediaType.Subsegment(startOfSubType);
+                }
+                return _mediaType.Subsegment(startOfSubType, startOfSuffix - startOfSubType);
             }
         }
 
         /// <summary>
-        /// SubType = "*"
+        /// Gets the structured syntax suffix of the <see cref="MediaTypeHeaderValue"/> if it has one.
+        /// See <see href="https://tools.ietf.org/html/rfc6838#section-4.8">The RFC documentation on structured syntaxes.</see>
         /// </summary>
-        public bool MatchesAllSubTypes
+        /// <example>
+        /// For the media type <c>"application/vnd.example+json"</c>, the property gives the value
+        /// <c>"json"</c>.
+        /// </example>
+        public StringSegment SubTypeSuffix
         {
             get
             {
-                return SubType.Equals("*", StringComparison.Ordinal);
+                var startOfSuffix = _mediaType.LastIndexOf(PlusCharacter);
```

Comment:
I'm just saying the indexing is wrong. You should be looking for the last index in the SubType.